### ISSUE TITLE
White flash problem

### ISSUE THREAD

**Issue by Enlight432** [jimschubert/NewTab-Redirect#228]
My entire work environment is in dark mode
When I open the tab, I encounter a white flash
Any help is appreciated

**Comment by mkbhru** [comment]
yes same here. Please admin atleast provide some functionality to have have a dark css while redirecting to some website in the new tab

**Comment by HonoredShadow** [comment]
I would like to request the removal of the white flash. I appreciate this extension, but I suffer from eye pain and headaches and must use low brightness settings. As mkbhru mentioned, could you please incorporate a feature for complete dark mode?

I'm grateful for this extension. Thank you.

**Comment by jimschubert** [comment]
I doubt there's a way to solve this, because the only thing the extension does is redirect which occurs before any CSS rendering. I'll play around with the prefers-color-scheme media annotation, but I'm not hopeful it'll help.

**Comment by JCollins2048** [comment]
That's all we can ask of you, Mr. Schubert.  Thank you.

**Comment by chriscorbell** [comment]
Enabling **chrome://flags/#enable-force-dark** alleviated this issue for me, with obvious caveats if you are willing to live with them

Edit: I ended up reverting this as it broke a few sites I visit regularly.

**Comment by Leftium** [comment]
> I doubt there's a way to solve this, because the only thing the extension does is redirect which occurs before any CSS rendering.

@jimschubert It *is* possible to eliminate/reduce the white flash:
- I made a simple POC by modifying the Blank New Tab example: https://github.com/Leftium/chrome-extensions-samples/commit/9b316d25116b7063d84eec2d431207f93851cafa.
- The POC simply styles the extension's `newtab`  override page with a dark background color, then adds a redirect via JS.
- I compared this POC against NewTab-Redirect:
  - The POC shows no flash when opening a new window; a slight gray "blink" when opening a new tab
  - NewTab-Redirect flashes a markedly longer/brighter white in both cases.
- Kagi loads very quickly; if the target site takes more time to load (like https://multi-launch.leftium.com) then the difference in white flash is even greater! (I can read the "Blank New Tab™" text before the redirect completes.)

I believe NewTab-Redirect loads CSS before the redirect; specifically: https://github.com/jimschubert/NewTab-Redirect/blob/422858b1b728c14e892f69faa7ba1dd76c74d1ad/main.html#L8-L9

Of course, my POC flashes black when in light mode, so `prefers-color-scheme` should be added.

(Platform is MacOS, if that matters.)

**Comment by Leftium** [comment]
**By moving the redirect script to the `<head>`, I think all white/black flashes were completely eliminated:**
- https://github.com/Leftium/chrome-extensions-samples/commit/08e04937b94152ac407990340821c569ec63c013
- The viewport area just takes the color of the tab chrome (black, white, purple... depending on the theme) until the fully rendered page appears.
- No flash of white or black...
- Not even a gray blink!

The full HTML for the replacement page looks like this:

```
<html>
  <head>
    <script src="redirect.js"></script> 
  </head>
</html>
```

**Comment by jimschubert** [comment]
redirect.js _has_ to be loaded last, that's how the extension works.

Google requires that every extension have a very clear entrypoint for users, meaning either a page or an extension icon with functionality aligning to the extension's behavior. Obviously I can't have an icon in the browser bar with "new tab" visuals, so I have to have a new tab page with visuals. If I were to redirect before those visuals I wouldn't be able to call `angular.resumeBootstrap` appropriately within the IIFE. I could move away from angular, but that would just mean I have to completely redo the UI.

If you _really_ want a performant redirection, don't even use javascript. Just create a local extension with contents like this:

```
<head>
    <meta http-equiv="refresh" content="0;URL=https://duckduckgo.com/">
</head>
```

**Comment by Leftium** [comment]
@jimschubert 
> redirect.js _has_ to be loaded last, that's how the extension works.

**I think it is possible to front-load just a small part of NewTab-Redirect's redirect.js** with minimal changes to the rest of the extension:
1. First, conditionally redirect in the `<head>` based on url/localStorage. (This doesn't seem to require angular at all.)
2. Then if not redirected, the angular portion of redirect.js can load. (Show/render the extension page.)

**Right now NewTab-Redirect is using this order:**
1. Initialize angular (Show/render the extension page.)
2. Conditionally redirect based on URL/localstorage.

**So this is the only part of redirect.js that would be loaded in the `<head>`:**  (The conditional redirect)
https://github.com/jimschubert/NewTab-Redirect/blob/422858b1b728c14e892f69faa7ba1dd76c74d1ad/js/redirect.js#L5-L10

---

Now, I am not sure if all the checks can be done from the head (Is `chrome.storage.local.get` available?) And perhaps redirecting from the head would violate some Chrome store rule?

So even if the JS logic is left the same, I think **updating the CSS to honor the `prefers-color-scheme` would mitigate the magnitude of the white flash.** Because CSS *is* being loaded and rendered:

https://github.com/jimschubert/NewTab-Redirect/blob/422858b1b728c14e892f69faa7ba1dd76c74d1ad/main.html#L8-L9

**Comment by Alpha12o** [comment]
Hello, any updates on this issue? There has to be some way to remove the white flash before loading the target URL.
Thanks
